lib/internal/modules/run_main.js

```diff
@@ -90,10 +92,16 @@ async function handleMainPromise(promise) {
  * Parse the CLI main entry point string and run it.
  * For backwards compatibility, we have to run a bunch of monkey-patchable code that belongs to the CJS loader (exposed
  * by `require('module')`) even when the entry point is ESM.
+ * This monkey-patchable code is bypassed under `--experimental-default-type=module`.
  * Because of backwards compatibility, this function is exposed publicly via `import { runMain } from 'node:module'`.
  * @param {string} main - Resolved absolute path for the main entry point, if found
  */
 function executeUserEntryPoint(main = process.argv[1]) {
+  if (getOptionValue('--experimental-default-type') === 'module') {
+    runMainESM(main);
+    return;
+  }
```

Comment:
Do we want to use the ESM loader to load `.cjs` file? After all why not, it makes it consistent, I like it, but we might want to update the documentation.

Reply:
I think I need to undo some of the latest commit though, because we _do_ need to resolve the entry to a path before converting it into a URL. I feel like that should be two steps, not just one where we pass a relative path into `import` that treats it as a URL; that’s just wrong, that can work most of the time but then fail for edge cases.

Reply:
But `pathToFileURL` already resolves it, no what you are describing is not happening. Or do you mean something else?

Reply:
> But `pathToFileURL` already resolves it

How? Are you referring to this `pathToFileURL`?

```js
    const main = path.isAbsolute(mainPath) ?
      pathToFileURL(mainPath).href : mainPath;
```

This `pathToFileURL` only gets called if `mainPath` is already an absolute path. Which it would be if we don’t skip or bail early from `resolveMainPath`, but that’s not what we’re doing in the latest version of this PR.

Reply:
Let's remove the condition and always pass it to `pathToFileURL` no? When would it not be an absolute path if you resolve it?

Reply:
> When would it not be an absolute path if you resolve it?

When it's not a path?

Reply:
What is the scenario where `mainPath` would not be a path?

Reply:
I don’t know. I just removed this and the tests still pass, so I guess we don’t need it. I also don’t like putting this inside the callback when it doesn’t need to be there. Pushed a commit.

Reply:
Yeah my impression was that code was written with assumptions that no longer apply, and it seems to me safe to delete.